### PR TITLE
[1/N] Add ASP.NET Support - Refactor Installation Worker

### PR DESCRIPTION
**Pull request by nagilson**
The first PR of many for adding support for ASP.NET Runtime installation.
I will try to break it up into smaller pieces.

This is a refactoring to make it so we don't need to write ourselves out of a paper bag when adding support for ASP.NET.
The `DotnetCoreAcquisitionWorker` is the main driving force for all installations.

Currently, the requested installation information is held in a `Context`. The `Context` used to be passed to the worker upon construction. This is an anti-pattern, though I can see several reasons why it was designed this way (the context object needs to be around in certain places beforehand for information to be accurate, coupled logic that needed to be decoupled, etc) but it ends up creating multiple sources of truth and requires several workers to be made which duplicates code.

Now the `Context` is passed to the worker per install request, which matches its real use case, considering the Context holds information about an individual installation request.

Without this change we will end up with a ton of code duplication and overhead to maintain a separate code path for ASP.NET management.

### CODE REVIEW

File: vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts [hunk 1]
```diff
@@ -262,87 +265,103 @@ export class DotnetCoreAcquisitionWorker implements IDotnetCoreAcquisitionWorker
...
-            this.context.eventStream.post(new DotnetAcquisitionPartialInstallation(installKey));
+            context.eventStream.post(new DotnetAcquisitionPartialInstallation(installKey));
 
             // Delete the existing local files so we can re-install. For global installs, let the installer handle it.
-            await this.uninstallLocalRuntimeOrSDK(installKey);
+            await this.uninstallLocalRuntimeOrSDK(context, installKey);
+        }
+    }
+
+    public async tryCleanUpInstallGraveyard(context: IAcquisitionWorkerContext) : Promise<void>
```
Comment (MiYanni) [inline vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts:342]: Does this graveyard contain... g.. g... g.... ghosts?!?! 👻😱

File: vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts [hunk 2]
```diff
@@ -262,87 +265,103 @@ export class DotnetCoreAcquisitionWorker implements IDotnetCoreAcquisitionWorker
...
+    {
+        const graveyard = new InstallationGraveyard(context);
+        const installsToRemove = await graveyard.get();
+        for(const install of installsToRemove)
+        {
+            context.eventStream.post(new DotnetInstallGraveyardEvent(
+                `Attempting to remove .NET at ${JSON.stringify(install)} again, as it was left in the graveyard.`));
```
Comment (MiYanni) [inline vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts:349]: I know the feeling of being left in the graveyard. 💀😭

File: vscode-dotnet-runtime-library/src/Acquisition/InstallTrackerSingleton.ts [hunk 3]
```diff
@@ -38,15 +47,33 @@ interface InProgressInstall
...
+    {
+
+    }
+
+    public static getInstance(eventStream : IEventStream, extensionState : IExtensionState) : InstallTrackerSingleton
     {
+        if(!InstallTrackerSingleton.instance)
+            {
+                InstallTrackerSingleton.instance = new InstallTrackerSingleton(eventStream, extensionState);
+            }
```
Comment (MiYanni) [inline vscode-dotnet-runtime-library/src/Acquisition/InstallTrackerSingleton.ts:68]: nit: Indentation.

**Review by MiYanni — APPROVED** [review]

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 4]
```diff
@@ -313,23 +304,40 @@ export function activate(context: vscode.ExtensionContext, extensionContext?: IE
...
-        await callWithErrorHandling(() => runtimeAcquisitionWorker.uninstallAll(), runtimeIssueContextFunctor(commandContext ? commandContext.errorConfiguration : undefined, 'uninstallAll'));
+        await callWithErrorHandling(async () =>
+        {
+            const mode = 'runtime' as DotnetInstallMode;
+            const worker = getAcquisitionWorker();
+            const installDirectoryProvider = directoryProviderFactory(mode, context.globalStoragePath);
+
+            await worker.uninstallAll(globalEventStream, installDirectoryProvider.getStoragePath(), context.globalState);
+        },
+            getIssueContext(existingPathConfigWorker)(commandContext ? commandContext.errorConfiguration : undefined, 'uninstallAll')
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:332]: Honestly I prefer it to the original... it's easier to read, but a bit more confusing what its doing at first, perhaps

File: vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts (hunk 1 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts:342]: Ghosts of .NET Installs, yes 😄 Maybe the WIX Ghost, too

File: vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts (hunk 2 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Acquisition/DotnetCoreAcquisitionWorker.ts:349]: I love this comment, thank you 🤣

File: vscode-dotnet-runtime-library/src/Acquisition/InstallTrackerSingleton.ts (hunk 3 above)
Comment (nagilson) [reply vscode-dotnet-runtime-library/src/Acquisition/InstallTrackerSingleton.ts:68]: ack I dont know why my auto indentation tool is trying to indent like that... thanks!

Comment (nagilson) [inline vscode-dotnet-runtime-library/src/Acquisition/InstallTrackerSingleton.ts:68]:
```suggestion
        {
            InstallTrackerSingleton.instance = new InstallTrackerSingleton(eventStream, extensionState);
        }
```

File: vscode-dotnet-runtime-extension/src/extension.ts [hunk 5]
```diff
@@ -178,20 +172,16 @@ export function activate(context: vscode.ExtensionContext, extensionContext?: IE
...
-            }
+            // Note: This will impact the context object given to the worker and error handler since objects own a copy of a reference in JS.
+            const runtimeVersionResolver = new VersionResolver(runtimeContext);
+            commandContext.version = await runtimeVersionResolver.getFullRuntimeVersion(commandContext.version);
 
             const acquisitionInvoker = new AcquisitionInvoker(runtimeContext, utilContext);
-            return runtimeAcquisitionWorker.acquireRuntime(version, acquisitionInvoker);
-        }, runtimeIssueContextFunctor(commandContext.errorConfiguration, 'acquire', commandContext.version), commandContext.requestingExtensionId, runtimeContext);
+            return worker.acquireRuntime(runtimeContext, acquisitionInvoker);
+        }, getIssueContext(existingPathConfigWorker)(commandContext.errorConfiguration, 'acquire', commandContext.version), commandContext.requestingExtensionId, runtimeContext);
```
[the remark this answers is not in the document]
Comment (nagilson) [reply vscode-dotnet-runtime-extension/src/extension.ts:181]: Yeah I don't know why it's like that either, the code was always like that there....  I dont think it should affect the way the code works at all lol. Maybe it was done that way to like, prevent less parameters going into 1 function? But it's kinda confusing. I was tempted to change it during the refactor but I decided it was not worth it and wanted to limit the scope of the refactor to what was necessary

**Pull request merged**